pytradfri/api/aiocoap_api.py

```diff
@@ -7,75 +7,51 @@
 from aiocoap import Message, Context
 from aiocoap.error import RequestTimedOut, Error, ConstructionRenderableError
 from aiocoap.numbers.codes import Code
-from aiocoap.transports import tinydtls
 
 from ..error import ClientError, ServerError, RequestTimeout
 from ..gateway import Gateway
 
 _LOGGER = logging.getLogger(__name__)
-
-
-class PatchedDTLSSecurityStore:
-    """Patched DTLS store in lieu of a credentials framework.
-       https://github.com/chrysn/aiocoap/issues/97"""
-
-    IDENTITY = None
-    KEY = None
-
-    def _get_psk(self, host, port):
-        return PatchedDTLSSecurityStore.IDENTITY, PatchedDTLSSecurityStore.KEY
-
-
-tinydtls.DTLSSecurityStore = PatchedDTLSSecurityStore
+_SENTINEL = object()
 
 
 class APIFactory:
-    def __init__(self, host, psk_id='pytradfri', psk=None, loop=None):
+    def __init__(self, host: str, psk_id='pytradfri', psk=None,
+                 internal_create=None):
+        if internal_create is not _SENTINEL:
+            raise ValueError("Use APIFactory.init(…) to initialize APIFactory")
+
         self._psk = psk
         self._host = host
         self._psk_id = psk_id
-        self._loop = loop
         self._observations_err_callbacks = []
         self._protocol = None
-
-        if self._loop is None:
-            self._loop = asyncio.get_event_loop()
-
         self._reset_lock = asyncio.Lock()
 
-        PatchedDTLSSecurityStore.IDENTITY = self._psk_id.encode('utf-8')
-
-        if self._psk:
-            PatchedDTLSSecurityStore.KEY = self._psk.encode('utf-8')
+    @classmethod
+    async def init(cls, host, psk_id='pytradfri', psk=None) -> "APIFactory":
+        """Initialize an APIFactory."""
+        instance = cls(host, psk_id=psk_id, psk=psk, internal_create=_SENTINEL)
+        if psk:
+            await instance._update_credentials()
+        return instance
 
     @property
     def psk_id(self):
         return self._psk_id
 
-    @psk_id.setter
-    def psk_id(self, value):
-        self._psk_id = value
-        PatchedDTLSSecurityStore.IDENTITY = self._psk_id.encode('utf-8')
-
     @property
     def psk(self):
         return self._psk
 
-    @psk.setter
-    def psk(self, value):
-        self._psk = value
-        PatchedDTLSSecurityStore.KEY = self._psk.encode('utf-8')
-
     async def _get_protocol(self):
         """Get the protocol for the request."""
         if self._protocol is None:
-            self._protocol = asyncio.Task(Context.create_client_context(
-                loop=self._loop))
+            self._protocol = asyncio.Task(Context.create_client_context())
```

Comment:
Probably use `asyncio.create_task` instead. We await the task below anyway so it should be the same result.

Reply:
Or do we need to support older Python versions than Python 3.7?

Reply:
If so the recommended way is to use `asyncio.ensure_future`.